## 1. Install required Libraries

In [2]:
!pip install numpy


                                              0.0/15.8 MB ? eta -:--:--
                                              0.1/15.8 MB 2.3 MB/s eta 0:00:07
                                              0.3/15.8 MB 3.4 MB/s eta 0:00:05
     -                                        0.5/15.8 MB 3.7 MB/s eta 0:00:05
     -                                        0.6/15.8 MB 3.6 MB/s eta 0:00:05
     --                                       0.9/15.8 MB 3.7 MB/s eta 0:00:05
     --                                       1.1/15.8 MB 3.8 MB/s eta 0:00:04
     ---                                      1.3/15.8 MB 4.0 MB/s eta 0:00:04
     ---                                      1.5/15.8 MB 4.2 MB/s eta 0:00:04
     ----                                     1.7/15.8 MB 4.2 MB/s eta 0:00:04
     -----                                    2.0/15.8 MB 4.4 MB/s eta 0:00:04
     -----                                    2.3/15.8 MB 4.6 MB/s eta 0:00:03
     ------                                   2.5/15.8 MB 4

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
altair 5.3.0 requires pandas>=0.25, which is not installed.
altair 5.3.0 requires toolz, which is not installed.
bayesian-optimization 1.4.1 requires scipy>=1.0.0, which is not installed.
category-encoders 2.6.3 requires pandas>=1.0.5, which is not installed.
category-encoders 2.6.3 requires patsy>=0.5.1, which is not installed.
category-encoders 2.6.3 requires scipy>=1.0.0, which is not installed.
category-encoders 2.6.3 requires statsmodels>=0.9.0, which is not installed.
feature-engine 1.6.1 requires pandas>=1.0.3, which is not installed.
feature-engine 1.6.1 requires scipy>=1.4.1, which is not installed.
feature-engine 1.6.1 requires statsmodels>=0.11.1, which is not installed.
fuzzytm 2.0.5 requires pandas, which is not installed.
fuzzytm 2.0.5 requires scipy, which is not installed.
imagehash 4.3.1 requires 

In [3]:
!pip install pandas

                                              0.0/11.6 MB ? eta -:--:--
                                              0.1/11.6 MB 1.7 MB/s eta 0:00:08
                                              0.2/11.6 MB 2.0 MB/s eta 0:00:06
                                              0.3/11.6 MB 2.0 MB/s eta 0:00:06
     -                                        0.4/11.6 MB 2.0 MB/s eta 0:00:06
     -                                        0.5/11.6 MB 2.0 MB/s eta 0:00:06
     -                                        0.6/11.6 MB 2.2 MB/s eta 0:00:06
     --                                       0.7/11.6 MB 2.1 MB/s eta 0:00:06
     --                                       0.8/11.6 MB 2.1 MB/s eta 0:00:06
     ---                                      1.0/11.6 MB 2.3 MB/s eta 0:00:05
     ---                                      1.0/11.6 MB 2.3 MB/s eta 0:00:05
     ---                                      1.1/11.6 MB 2.2 MB/s eta 0:00:05
     ----                                     1.3/11.6 MB 2

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
altair 5.3.0 requires toolz, which is not installed.
category-encoders 2.6.3 requires patsy>=0.5.1, which is not installed.
category-encoders 2.6.3 requires scipy>=1.0.0, which is not installed.
category-encoders 2.6.3 requires statsmodels>=0.9.0, which is not installed.
feature-engine 1.6.1 requires scipy>=1.4.1, which is not installed.
feature-engine 1.6.1 requires statsmodels>=0.11.1, which is not installed.
fuzzytm 2.0.5 requires scipy, which is not installed.
mlxtend 0.23.0 requires matplotlib>=3.0.0, which is not installed.
mlxtend 0.23.0 requires scipy>=1.2.1, which is not installed.
pandas-profiling 3.2.0 requires matplotlib>=3.2.0, which is not installed.
pandas-profiling 3.2.0 requires scipy>=1.4.1, which is not installed.
pandas-profiling 3.2.0 requires seaborn>=0.10.1, which is not installed.
pandas-pr

In [5]:
!pip install scikit-learn

                                              0.0/46.2 MB ? eta -:--:--
                                              0.1/46.2 MB 1.3 MB/s eta 0:00:35
                                              0.2/46.2 MB 2.8 MB/s eta 0:00:17
                                              0.4/46.2 MB 3.6 MB/s eta 0:00:13
                                              0.7/46.2 MB 3.8 MB/s eta 0:00:13
                                              0.9/46.2 MB 4.5 MB/s eta 0:00:10
                                              1.0/46.2 MB 4.0 MB/s eta 0:00:12
     -                                        1.4/46.2 MB 4.5 MB/s eta 0:00:10
     -                                        1.7/46.2 MB 4.8 MB/s eta 0:00:10
     -                                        2.0/46.2 MB 5.0 MB/s eta 0:00:09
     -                                        2.2/46.2 MB 5.1 MB/s eta 0:00:09
     --                                       2.5/46.2 MB 5.0 MB/s eta 0:00:09
     --                                       2.8/46.2 MB 5

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
category-encoders 2.6.3 requires patsy>=0.5.1, which is not installed.
category-encoders 2.6.3 requires statsmodels>=0.9.0, which is not installed.
feature-engine 1.6.1 requires statsmodels>=0.11.1, which is not installed.
imagehash 4.3.1 requires pillow, which is not installed.
imagehash 4.3.1 requires PyWavelets, which is not installed.
missingno 0.5.2 requires matplotlib, which is not installed.
missingno 0.5.2 requires seaborn, which is not installed.
mlxtend 0.23.0 requires matplotlib>=3.0.0, which is not installed.
pandas-profiling 3.2.0 requires matplotlib>=3.2.0, which is not installed.
pandas-profiling 3.2.0 requires seaborn>=0.10.1, which is not installed.
pandas-profiling 3.2.0 requires tqdm>=4.48.2, which is not installed.
phik 0.12.4 requires matplotlib>=2.2.3, which is not installed.
pmdarima 2.0.4 r

## 2. Importing Libraries

In [6]:
import os

import numpy as np

import pandas as pd

from sklearn.model_selection import train_test_split

## 3. Reading the Data

In [146]:
PROJECT_DIR = r"E:\sagemaker-flight-prices-prediction"
DATA = "data"
DATA_DIR = "data"

In [10]:
def get_data(name):
    file_name = f"{name}.csv"
    file_path = os.path.join(PROJECT_DIR, DATA, file_name)
    return pd.read_csv(file_path)

In [12]:
flights = get_data('flight_price')

In [13]:
flights.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [14]:
flights

Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [15]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


- The dataset contains 10683 rows with 11 features
- Feature Route & Total_Stops have one missing each
- Data types of features Date_of_Journey, Dep_Time, Arrival_Time, Duration are not appropriate

## 4. Preliminary Analysis

## 4.1 Check Data types

In [16]:
flights.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

In [37]:
flights.columns.to_list()

['Airline',
 'Date_of_Journey',
 'Source',
 'Destination',
 'Route',
 'Dep_Time',
 'Arrival_Time',
 'Duration',
 'Total_Stops',
 'Additional_Info',
 'Price']

In [17]:
flights.Date_of_Journey

0        24/03/2019
1         1/05/2019
2         9/06/2019
3        12/05/2019
4        01/03/2019
            ...    
10678     9/04/2019
10679    27/04/2019
10680    27/04/2019
10681    01/03/2019
10682     9/05/2019
Name: Date_of_Journey, Length: 10683, dtype: object

In [18]:
flights.Date_of_Journey.iloc[6]  # The feature is a string. Data type to be changed.

'12/03/2019'

In [20]:
flights.Dep_Time.iloc[10]  # The feature is a string. Data type to be changed.

'09:45'

In [21]:
flights.Duration.iloc[6] # The feature is a string. Data type to be changed. Mix of hrs and minutes

'15h 30m'

In [22]:
flights.Airline.unique()  # Airline name and seat class mixed up

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [25]:
flights.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

In [26]:
flights.Destination.unique() # Delhi & New Delhi Same

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

In [27]:
flights.Total_Stops.unique() # Data type inappropriate. Can replace with integer.

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

## 4.2 Check Data types

In [33]:
flights.duplicated(keep=False)

0        False
1        False
2        False
3        False
4        False
         ...  
10678    False
10679    False
10680    False
10681    False
10682    False
Length: 10683, dtype: bool

In [29]:
flights.duplicated().sum()  # 220 duplicate rows present in dataset

220

In [45]:
# Look at duplicate data
(
    flights
    .loc[flights.duplicated(keep='first')]
    .sort_values(['Airline', 'Date_of_Journey', 'Source', 'Destination'])

)

Airline Date_of_Journey    Source Destination  \
9848    Air India      01/03/2019  Banglore   New Delhi   
8168    Air India      03/03/2019  Banglore   New Delhi   
9913    Air India       1/04/2019   Kolkata    Banglore   
5042    Air India       1/05/2019   Kolkata    Banglore   
6377    Air India       1/05/2019   Kolkata    Banglore   
...           ...             ...       ...         ...   
9216  Jet Airways       9/06/2019     Delhi      Cochin   
7502     SpiceJet      03/03/2019  Banglore   New Delhi   
2870     SpiceJet      24/03/2019  Banglore   New Delhi   
3711     SpiceJet      24/03/2019  Banglore   New Delhi   
5470      Vistara      24/03/2019  Banglore   New Delhi   

                      Route Dep_Time  Arrival_Time Duration Total_Stops  \
9848  BLR → BOM → AMD → DEL    08:50  23:55 02 Mar   39h 5m     2 stops   
8168              BLR → DEL    21:10         23:55   2h 45m    non-stop   
9913  CCU → DEL → COK → BLR    10:00  01:20 02 Apr  15h 20m     2 stops   
5042  CCU → DEL → COK → BLR    10:00  13:45 02 May  27h 45m     2 stops   
6377  CCU → DEL → COK → BLR    10:00  13:45 02 May  27h 45m     2 stops   
...                     ...      ...           ...      ...         ...   
9216  DEL → IDR → BOM → COK    05:30  04:25 10 Jun  22h 55m     2 stops   
7502              BLR → DEL    20:30         23:20   2h 50m    non-stop   
2870              BLR → DEL    05:45         08:35   2h 50m    non-stop   
3711              BLR → DEL    20:30         23:20   2h 50m    non-stop   
5470              BLR → DEL    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
9848                       No info  17135  
8168                       No info   7591  
9913                       No info  10408  
5042                       No info  15164  
6377                       No info  15164  
...                            ...    ...  
9216                       No info  13292  
7502                       No info   6860  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
5470                       No info   5403  

[220 rows x 11 columns]

In [46]:
(
    flights
    .loc[flights.duplicated(keep=False)]
    .sort_values(['Airline', 'Date_of_Journey', 'Source', 'Destination'])

)

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

## 4.3 Observations
- There're 220 duplicates. These should be removed.
- The type of Duration and Total_Stops is mixed. It should be numeric type.
- The type of Date_of_Journey, Dep_Time and Arrival_Time should be changed to datetime

## 5. Detailed Analysis 

# Airline

In [47]:
flights.columns.to_list()

['Airline',
 'Date_of_Journey',
 'Source',
 'Destination',
 'Route',
 'Dep_Time',
 'Arrival_Time',
 'Duration',
 'Total_Stops',
 'Additional_Info',
 'Price']

In [48]:
flights.Airline

0             IndiGo
1          Air India
2        Jet Airways
3             IndiGo
4             IndiGo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

In [49]:
flights.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [50]:
flights.Airline.str

In [51]:
(
    flights
    .Airline
    .str.replace('Premium economy', '')
    .str.replace('Business', '')
    .str.title()


)

0             Indigo
1          Air India
2        Jet Airways
3             Indigo
4             Indigo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

In [52]:
flights.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

# Date_of_Journey

In [54]:
flights.Date_of_Journey

0        24/03/2019
1         1/05/2019
2         9/06/2019
3        12/05/2019
4        01/03/2019
            ...    
10678     9/04/2019
10679    27/04/2019
10680    27/04/2019
10681    01/03/2019
10682     9/05/2019
Name: Date_of_Journey, Length: 10683, dtype: object

In [57]:
pd.to_datetime(flights.Date_of_Journey, dayfirst=True)

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

# Source

In [58]:
flights.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

# Destination

In [59]:
flights.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

In [63]:
(
    flights
    .Destination
    .str.replace('New Delhi', 'Delhi')
    
)

0           Delhi
1        Banglore
2          Cochin
3        Banglore
4           Delhi
           ...   
10678    Banglore
10679    Banglore
10680       Delhi
10681       Delhi
10682      Cochin
Name: Destination, Length: 10683, dtype: object

0           Delhi
1        Banglore
2          Cochin
3        Banglore
4           Delhi
           ...   
10678    Banglore
10679    Banglore
10680       Delhi
10681       Delhi
10682      Cochin
Name: Destination, Length: 10683, dtype: object

# Route

- Route and Total_Stops represent same information. Route feature can be dropped.

# Dep_Time

In [64]:
flights.Dep_Time

0        22:20
1        05:50
2        09:25
3        18:05
4        16:50
         ...  
10678    19:55
10679    20:45
10680    08:20
10681    11:30
10682    10:55
Name: Dep_Time, Length: 10683, dtype: object

In [68]:
(
    flights
    .Dep_Time
    .loc[lambda ser: ser.str.contains("[^0-9:]")]
    
)

Series([], Name: Dep_Time, dtype: object)

In [70]:
pd.to_datetime(flights.Dep_Time).dt.time

C:\Users\akellavp\AppData\Local\Temp\ipykernel_4908\1056054517.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(flights.Dep_Time).dt.time


0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: Dep_Time, Length: 10683, dtype: object

# Arrival_Time

In [71]:
flights.Arrival_Time

0        01:10 22 Mar
1               13:15
2        04:25 10 Jun
3               23:30
4               21:35
             ...     
10678           22:25
10679           23:20
10680           11:20
10681           14:10
10682           19:15
Name: Arrival_Time, Length: 10683, dtype: object

In [73]:
flights.Arrival_Time.str.split(pat=" ")

0        [01:10, 22, Mar]
1                 [13:15]
2        [04:25, 10, Jun]
3                 [23:30]
4                 [21:35]
               ...       
10678             [22:25]
10679             [23:20]
10680             [11:20]
10681             [14:10]
10682             [19:15]
Name: Arrival_Time, Length: 10683, dtype: object

In [82]:
(
	flights
	.Arrival_Time
	.loc[lambda ser: ser.str.contains("[^0-9:]")]
	.str.split(" ", n=1)
	.str.get(1)
	.unique()
)

array(['22 Mar', '10 Jun', '13 Mar', '02 Mar', '10 May', '04 Mar',
       '13 Jun', '28 May', '19 Mar', '07 May', '02 Jun', '16 Jun',
       '19 May', '16 May', '28 Jun', '02 May', '28 Mar', '19 Jun',
       '04 Apr', '25 Mar', '07 Mar', '25 Jun', '07 Jun', '25 May',
       '13 May', '16 Mar', '22 May', '10 Apr', '04 Jun', '20 May',
       '28 Apr', '25 Apr', '10 Mar', '19 Apr', '13 Apr', '02 Apr',
       '23 Mar', '22 Apr', '11 May', '07 Apr', '03 May', '08 Mar',
       '03 Mar', '05 Mar', '22 Jun', '04 May', '26 May', '16 Apr',
       '26 Jun', '29 May', '29 Jun', '29 Mar', '23 May', '17 Jun'],
      dtype=object)

# Duration

In [83]:
flights.Duration

0        2h 50m
1        7h 25m
2           19h
3        5h 25m
4        4h 45m
          ...  
10678    2h 30m
10679    2h 35m
10680        3h
10681    2h 40m
10682    8h 20m
Name: Duration, Length: 10683, dtype: object

In [85]:
# Records with Duration in only hours
(
	flights
	.Duration
	.loc[lambda ser: ~ser.str.contains("m")]
	# .str.split(" ", n=1)
	# .str.get(1)
	 .unique()
)

array(['19h', '23h', '22h', '12h', '3h', '5h', '10h', '18h', '24h', '15h',
       '16h', '8h', '14h', '20h', '13h', '11h', '9h', '27h', '26h', '4h',
       '7h', '30h', '21h', '28h', '47h', '6h', '25h', '38h', '34h'],
      dtype=object)

In [86]:
# Records with Duration in only hours
(
	flights
	.Duration
	.loc[lambda ser: ~ser.str.contains("h")]
	# .str.split(" ", n=1)
	# .str.get(1)
	# .unique()
)

# 5m duration of flight is not possible. Drop this row.

6474    5m
Name: Duration, dtype: object

In [90]:
flights.iloc[[6474]]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

In [102]:
(
    flights
    .Duration
    .drop(index=[6474])
    .str.split(" ", expand=True) # Splits and creates multiple columns from one column
    .set_axis(['hour', 'minute'], axis = 1) # Sets column names
    .assign(                                # Strips h from hour, converts to int dtype and converts to minutes
            hour = lambda df_: (
                df_
                .hour
                .str.replace("h", "")
                .astype(int)
                .mul(60)
                
            ),

            minute = lambda df_: (
                df_
                .minute
                .str.replace("m", "")
                .fillna("0")   # When hr & minute are split some values in minute column are Nan
                .astype(int)
            )
    )
    .sum(axis=1)
    .rename('duration_minutes')
    .to_frame()
    .join(flights.Duration.drop(index=[6474])) # Join with original column for comparision
    
    
)

duration_minutes Duration
0                   170   2h 50m
1                   445   7h 25m
2                  1140      19h
3                   325   5h 25m
4                   285   4h 45m
...                 ...      ...
10678               150   2h 30m
10679               155   2h 35m
10680               180       3h
10681               160   2h 40m
10682               500   8h 20m

[10682 rows x 2 columns]

In [99]:
flights.Duration.head(10)

0     2h 50m
1     7h 25m
2        19h
3     5h 25m
4     4h 45m
5     2h 25m
6    15h 30m
7     21h 5m
8    25h 30m
9     7h 50m
Name: Duration, dtype: object

# Total_Stops

In [105]:
flights.Total_Stops

0        non-stop
1         2 stops
2         2 stops
3          1 stop
4          1 stop
           ...   
10678    non-stop
10679    non-stop
10680    non-stop
10681    non-stop
10682     2 stops
Name: Total_Stops, Length: 10683, dtype: object

In [106]:
flights.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [115]:
(
    flights
    .Total_Stops
    .replace('non-stop', "0")
    .str.replace(" stops?", "", regex=True)   # " stop" " stops"
    # .replace('nan', "0")
    # .astype(float)
    .pipe(lambda ser: pd.to_numeric(ser))
)

0        0.0
1        2.0
2        2.0
3        1.0
4        1.0
        ... 
10678    0.0
10679    0.0
10680    0.0
10681    0.0
10682    2.0
Name: Total_Stops, Length: 10683, dtype: float64

# Additional_Info

In [117]:
flights.Additional_Info.unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

## 6. Cleaning Operations

In [135]:
def convert_to_minutes(ser):
    return (
        
            ser
            .str.split(" ", expand=True) # Splits and creates multiple columns from one column
            .set_axis(['hour', 'minute'], axis = 1) # Sets column names
            .assign(                                # Strips h from hour, converts to int dtype and converts to minutes
                    hour = lambda df_: (
                        df_
                        .hour
                        .str.replace("h", "")
                        .astype(int)
                        .mul(60)
                        
                    ),
        
                    minute = lambda df_: (
                        df_
                        .minute
                        .str.replace("m", "")
                        .fillna("0")   # When hr & minute are split some values in minute column are Nan
                        .astype(int)
                    )
            )
            .sum(axis=1)

    )
    

In [136]:
def clean_data(df):
	return (
    		df
    		.drop(index=[6474])
    		.drop_duplicates()
    		.assign(**{
    			col: df[col].str.strip()
    			for col in df.select_dtypes(include="O").columns
    		})
    		.rename(columns=str.lower)
    		.assign(
    			airline=lambda df_: (
    				df_
    				.airline
    				.str.replace(" Premium economy", "")
    				.str.replace(" Business", "")
    				.str.title()
    			),
    			date_of_journey=lambda df_: pd.to_datetime(df_.date_of_journey, dayfirst=True),
                destination=lambda df_:
                                    df_
                                    .destination
                                    .str.replace('New Delhi', 'Delhi'),
    			dep_time=lambda df_: pd.to_datetime(df_.dep_time).dt.time,
    			arrival_time=lambda df_: pd.to_datetime(df_.arrival_time).dt.time,
    			duration=lambda df_: df_.duration.pipe(convert_to_minutes),
    			total_stops=lambda df_: (
    				df_
    				.total_stops
    				.replace("non-stop", "0")
    				.str.replace(" stops?", "", regex=True)
    				.pipe(lambda ser: pd.to_numeric(ser))
    			),
    			additional_info=lambda df_: df_.additional_info.replace("No info", "No Info")
    		)
    		.drop(columns="route")
	)

In [137]:
flights_cleaned = clean_data(flights)
flights_cleaned

C:\Users\akellavp\AppData\Local\Temp\ipykernel_4908\2580447345.py:24: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time=lambda df_: pd.to_datetime(df_.dep_time).dt.time,
C:\Users\akellavp\AppData\Local\Temp\ipykernel_4908\2580447345.py:25: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time=lambda df_: pd.to_datetime(df_.arrival_time).dt.time,


airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore       Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore       Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore       Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         No Info   3897  
1         13:15:00       445          2.0         No Info   7662  
2         04:25:00      1140          2.0         No Info  13882  
3         23:30:00       325          1.0         No Info   6218  
4         21:35:00       285          1.0         No Info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         No Info   4107  
10679     23:20:00       155          0.0         No Info   4145  
10680     11:20:00       180          0.0         No Info   7229  
10681     14:10:00       160          0.0         No Info  12648  
10682     19:15:00       500          2.0         No Info  11753  

[10462 rows x 10 columns]

In [138]:
flights_cleaned

airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore       Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore       Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore       Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         No Info   3897  
1         13:15:00       445          2.0         No Info   7662  
2         04:25:00      1140          2.0         No Info  13882  
3         23:30:00       325          1.0         No Info   6218  
4         21:35:00       285          1.0         No Info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         No Info   4107  
10679     23:20:00       155          0.0         No Info   4145  
10680     11:20:00       180          0.0         No Info   7229  
10681     14:10:00       160          0.0         No Info  12648  
10682     19:15:00       500          2.0         No Info  11753  

[10462 rows x 10 columns]

## 7. Split the Data

In [139]:
X = flights_cleaned.drop(columns=['price'])
X

airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore       Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore       Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore       Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  
0         01:10:00       170          0.0         No Info  
1         13:15:00       445          2.0         No Info  
2         04:25:00      1140          2.0         No Info  
3         23:30:00       325          1.0         No Info  
4         21:35:00       285          1.0         No Info  
...            ...       ...          ...             ...  
10678     22:25:00       150          0.0         No Info  
10679     23:20:00       155          0.0         No Info  
10680     11:20:00       180          0.0         No Info  
10681     14:10:00       160          0.0         No Info  
10682     19:15:00       500          2.0         No Info  

[10462 rows x 9 columns]

In [140]:
y = flights_cleaned['price']
y

0         3897
1         7662
2        13882
3         6218
4        13302
         ...  
10678     4107
10679     4145
10680     7229
10681    12648
10682    11753
Name: price, Length: 10462, dtype: int64

In [142]:
# Create train/test/validation datasets
X_, X_test, y_, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_, y_, test_size=0.2, random_state=42)

print(X_train.shape, y_train.shape)
print(X_val.shape, y_val.shape)
print(X_test.shape, y_test.shape)

(6695, 9) (6695,)
(1674, 9) (1674,)
(2093, 9) (2093,)


## 8. Export the Data Subsets

In [144]:
def export_data(X, y, name):
    file_name = f"{name}.csv"
    file_path = os.path.join(PROJECT_DIR, DATA_DIR, file_name)
    X.join(y).to_csv(file_path, index=False)
    return pd.read_csv(file_path).head()
    

In [143]:
X_train.join(y_train).to_csv(fil)

airline date_of_journey    source destination  dep_time  \
3635  Jet Airways      2019-03-21  Banglore       Delhi  08:55:00   
613   Jet Airways      2019-03-27     Delhi      Cochin  17:30:00   
7390        Goair      2019-03-09  Banglore       Delhi  11:40:00   
7575    Air India      2019-06-12   Kolkata    Banglore  09:25:00   
7032  Jet Airways      2019-03-12  Banglore       Delhi  22:55:00   
...           ...             ...       ...         ...       ...   
4128  Jet Airways      2019-03-21     Delhi      Cochin  10:45:00   
2836    Air India      2019-05-01   Kolkata    Banglore  09:25:00   
6664  Jet Airways      2019-06-01     Delhi      Cochin  14:00:00   
7999     Air Asia      2019-06-24     Delhi      Cochin  07:55:00   
6024    Air India      2019-03-01  Banglore       Delhi  11:50:00   

     arrival_time  duration  total_stops              additional_info  price  
3635     19:10:00       615          1.0  In-flight meal not included   7832  
613      04:25:00       655          1.0  In-flight meal not included   6540  
7390     14:35:00       175          0.0                      No Info   7305  
7575     18:30:00       545          1.0                      No Info   8366  
7032     07:40:00       525          1.0  In-flight meal not included  11087  
...           ...       ...          ...                          ...    ...  
4128     18:50:00      1925          2.0                      No Info  11093  
2836     18:30:00       545          1.0                      No Info   8891  
6664     19:00:00       300          1.0  In-flight meal not included  10262  
7999     13:25:00       330          1.0                      No Info   6152  
6024     08:55:00      1265          1.0               1 Long layover  14594  

[6695 rows x 10 columns]

In [147]:
export_data(X_train,y_train,"train")

airline date_of_journey    source destination  dep_time arrival_time  \
0  Jet Airways      2019-03-21  Banglore       Delhi  08:55:00     19:10:00   
1  Jet Airways      2019-03-27     Delhi      Cochin  17:30:00     04:25:00   
2        Goair      2019-03-09  Banglore       Delhi  11:40:00     14:35:00   
3    Air India      2019-06-12   Kolkata    Banglore  09:25:00     18:30:00   
4  Jet Airways      2019-03-12  Banglore       Delhi  22:55:00     07:40:00   

   duration  total_stops              additional_info  price  
0       615          1.0  In-flight meal not included   7832  
1       655          1.0  In-flight meal not included   6540  
2       175          0.0                      No Info   7305  
3       545          1.0                      No Info   8366  
4       525          1.0  In-flight meal not included  11087

In [148]:
export_data(X_test,y_test,"test")

airline date_of_journey    source destination  dep_time arrival_time  \
0  Jet Airways      2019-03-06  Banglore       Delhi  08:00:00     08:15:00   
1     Spicejet      2019-06-06   Kolkata    Banglore  22:20:00     00:40:00   
2       Indigo      2019-03-18   Kolkata    Banglore  05:30:00     08:20:00   
3       Indigo      2019-06-27   Chennai     Kolkata  19:35:00     21:55:00   
4       Indigo      2019-05-06   Kolkata    Banglore  15:15:00     17:45:00   

   duration  total_stops additional_info  price  
0      1455          1.0         No Info  17996  
1       140          0.0         No Info   3873  
2       170          0.0         No Info   4462  
3       140          0.0         No Info   3597  
4       150          0.0         No Info   4804

In [149]:
export_data(X_val,y_val,"val")

airline date_of_journey    source destination  dep_time  \
0             Indigo      2019-06-24     Delhi      Cochin  20:25:00   
1  Multiple Carriers      2019-06-12     Delhi      Cochin  09:45:00   
2        Jet Airways      2019-03-12  Banglore       Delhi  22:55:00   
3  Multiple Carriers      2019-06-06     Delhi      Cochin  13:00:00   
4        Jet Airways      2019-05-18     Delhi      Cochin  23:05:00   

  arrival_time  duration  total_stops              additional_info  price  
0     01:30:00       305          1.0                      No Info   5054  
1     22:30:00       765          1.0                      No Info   9646  
2     15:15:00       980          1.0  In-flight meal not included  11087  
3     21:00:00       480          1.0                      No Info  13587  
4     04:25:00      1760          2.0                      No Info  16704

In [159]:
os.chdir("E:\sagemaker-flight-prices-prediction\data")
test = pd.read_csv('test.csv')

In [160]:
test

airline date_of_journey    source destination  dep_time  \
0           Jet Airways      2019-03-06  Banglore       Delhi  08:00:00   
1              Spicejet      2019-06-06   Kolkata    Banglore  22:20:00   
2                Indigo      2019-03-18   Kolkata    Banglore  05:30:00   
3                Indigo      2019-06-27   Chennai     Kolkata  19:35:00   
4                Indigo      2019-05-06   Kolkata    Banglore  15:15:00   
...                 ...             ...       ...         ...       ...   
2088        Jet Airways      2019-05-27     Delhi      Cochin  19:15:00   
2089  Multiple Carriers      2019-06-27     Delhi      Cochin  11:25:00   
2090        Jet Airways      2019-06-03     Delhi      Cochin  02:15:00   
2091  Multiple Carriers      2019-06-06     Delhi      Cochin  15:15:00   
2092  Multiple Carriers      2019-06-03     Delhi      Cochin  18:15:00   

     arrival_time  duration  total_stops              additional_info  price  
0        08:15:00      1455          1.0                      No Info  17996  
1        00:40:00       140          0.0                      No Info   3873  
2        08:20:00       170          0.0                      No Info   4462  
3        21:55:00       140          0.0                      No Info   3597  
4        17:45:00       150          0.0                      No Info   4804  
...           ...       ...          ...                          ...    ...  
2088     12:35:00      1040          1.0  In-flight meal not included  12898  
2089     19:15:00       470          1.0                      No Info   7155  
2090     04:25:00      1570          1.0  In-flight meal not included  11627  
2091     01:30:00       615          1.0                      No Info   6795  
2092     01:30:00       435          1.0                      No Info  13377  

[2093 rows x 10 columns]

In [161]:
test.duplicated().sum()

0

In [ ]:
(
    flights
    .Airline
    .
)

In [162]:
!pip install jupyter_contrib_nbextensions

                                              0.0/23.5 MB ? eta -:--:--
                                              0.0/23.5 MB 2.0 MB/s eta 0:00:12
                                              0.2/23.5 MB 2.6 MB/s eta 0:00:09
                                              0.2/23.5 MB 1.8 MB/s eta 0:00:14
                                              0.4/23.5 MB 2.6 MB/s eta 0:00:10
                                              0.5/23.5 MB 2.6 MB/s eta 0:00:09
     -                                        0.7/23.5 MB 2.9 MB/s eta 0:00:08
     -                                        0.8/23.5 MB 2.8 MB/s eta 0:00:09
     -                                        1.1/23.5 MB 3.1 MB/s eta 0:00:08
     --                                       1.3/23.5 MB 3.3 MB/s eta 0:00:07
     --                                       1.4/23.5 MB 3.3 MB/s eta 0:00:07
     --                                       1.6/23.5 MB 3.3 MB/s eta 0:00:07
     ---                                      1.9/23.5 MB 3


[notice] A new release of pip is available: 23.1.2 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip
